### PR TITLE
Fix: [Network] connections can use an invalid socket due to a race condition

### LINKED ISSUE #9730: [Bug]: crash on opening Multiplayer
### Version of OpenTTD

12.1, FreeBSD

### Expected result

No crash

### Actual result

As reported in #9728 by @RealDeuce.

```
Thread 7 received signal SIGBUS, Bus error.
Address not present.
[Switching to LWP 100805 of process 32847]
0x0000000000bb44b4 in NetworkTCPSocketHandler::CanSendReceive (this=0x807fcc488) at /usr/ports/games/openttd/work/openttd-12.1/src/network/core/tcp.cpp:226
226		FD_SET(this->sock, &read_fd);
(gdb) print this->sock
$1 = -1
(gdb) bt
#0  0x0000000000bb44b4 in NetworkTCPSocketHandler::CanSendReceive (this=0x807fcc488)
    at /usr/ports/games/openttd/work/openttd-12.1/src/network/core/tcp.cpp:226
#1  0x0000000000c318a6 in QueryNetworkGameSocketHandler::Receive (this=0x807fcc480)
    at /usr/ports/games/openttd/work/openttd-12.1/src/network/network_query.cpp:53
#2  0x0000000000c31ca1 in QueryNetworkGameSocketHandler::SendReceive ()
    at /usr/ports/games/openttd/work/openttd-12.1/src/network/network_query.cpp:147
#3  0x0000000000bd3f61 in NetworkBackgroundLoop ()
    at /usr/ports/games/openttd/work/openttd-12.1/src/network/network.cpp:1027
#4  0x000000000104e9d9 in GameLoop () at /usr/ports/games/openttd/work/openttd-12.1/src/openttd.cpp:1454
#5  0x0000000000da9d5e in VideoDriver::GameLoop (this=0x803694c00)
    at /usr/ports/games/openttd/work/openttd-12.1/src/video/video_driver.cpp:37
#6  0x0000000000da9dcb in VideoDriver::GameThread (this=0x803694c00)
    at /usr/ports/games/openttd/work/openttd-12.1/src/video/video_driver.cpp:44
#7  0x0000000000da9f05 in VideoDriver::GameThreadThunk (drv=0x803694c00)
    at /usr/ports/games/openttd/work/openttd-12.1/src/video/video_driver.cpp:81
#8  0x0000000000daba14 in StartNewThread<void (*)(VideoDriver*), VideoDriver*>(std::__1::thread*, char const*, void (*&&)(VideoDriver*), VideoDriver*&&)::{lambda(char const*, void (*&&)(VideoDriver*), VideoDriver*&&)#1}::operator()(char const*, void (*&&)(VideoDriver*), VideoDriver*&&) const (this=0x82ac66fa0, name=0x5e2bcd "ottd:game", 
    F=@0x82ac66fb0: 0xda9ef0 <VideoDriver::GameThreadThunk(VideoDriver*)>, A=@0x82ac66fb8: 0x803694c00)
    at /usr/ports/games/openttd/work/openttd-12.1/src/video/../thread.h:65
#9  0x0000000000dab8cf in std::__1::__invoke<StartNewThread<void (*)(VideoDriver*), VideoDriver*>(std::__1::thread*, char const*, void (*&&)(VideoDriver*), VideoDriver*&&)::{lambda(char const*, void (*&&)(VideoDriver*), VideoDriver*&&)#1}, char const*, void (*)(VideoDriver*), VideoDriver*> (__f=..., __args=@0x82ac66fb8: 0x803694c00, 
    __args=@0x82ac66fb8: 0x803694c00, __args=@0x82ac66fb8: 0x803694c00) at /usr/include/c++/v1/type_traits:3539
#10 0x0000000000dab830 in std::__1::__thread_execute<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, StartNewThread<void (*)(VideoDriver*), VideoDriver*>(std::__1::thread*, char const*, void (*&&)(VideoDriver*), VideoDriver*&&)::{lambda(char const*, void (*&&)(VideoDriver*), VideoDriver*&&)#1}, char const*, void (*)(VideoDriver*), VideoDriver*, 2ul, 3ul, 4ul>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, StartNewThread<void (*)(VideoDriver*), VideoDriver*>(std::__1::thread*, char const*, void (*&&)(VideoDriver*), VideoDriver*&&)::{lambda(char const*, void (*&&)(VideoDriver*), VideoDriver*&&)#1}, char const*, void (*)(VideoDriver*), VideoDriver*>&, std::__1::__tuple_indices<2ul, 3ul, 4ul>)
    (__t=...) at /usr/include/c++/v1/thread:273
#11 0x0000000000dab486 in std::__1::__thread_proxy<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct> >, StartNewThread<void (*)(VideoDriver*), VideoDriver*>(std::__1::thread*, char const*, void (*&&)(VideoDriver*), VideoDriver*&&)::{lambda(char const*, void (*&&)(VideoDriver*), VideoDriver*&&)#1}, char const*, void (*)(VideoDriver*), VideoDriver*> >(void*) (__vp=0x82ac66fa0)
    at /usr/include/c++/v1/thread:284
#12 0x0000000801cbcfac in ?? () from /lib/libthr.so.3
#13 0x0000000000000000 in ?? ()
Backtrace stopped: Cannot access memory at address 0x7fffdf5f9000
```

With `-dnet=6`:

```
openttd -dnet=6
dbg: [net] Starting network
dbg: [net] Initializing UDP listeners
dbg: [net] Network online, multiplayer available
dbg: [net] Detected broadcast addresses:
dbg: [net]   0) 192.168.0.255
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
fluidsynth: error: fluid_is_soundfont(): fopen() failed: 'File does not exist.'
dbg: [net] 44.135.218.201:3979 resolved in:
dbg: [net] - 44.135.218.201:3979 (IPv4)
dbg: [net] Attempting to connect to 44.135.218.201:3979 (IPv4)
dbg: [net] dragon.rrx.ca:3979 resolved in:
dbg: [net] - [2602:80b:7007::65]:3979 (IPv6)
dbg: [net] - 44.31.37.65:3979 (IPv4)
dbg: [net] Attempting to connect to [2602:80b:7007::65]:3979 (IPv6)
dbg: [net] Attempting to connect to 44.31.37.65:3979 (IPv4)
dbg: [net] Could not connect to [2602:80b:7007::65]:3979 (IPv6): Connection refused
dbg: [net] Connected to dragon.rrx.ca:3979
dbg: [net] Failed to get address of the peer: Bad file descriptor
dbg: [net] - using [::]:0 (IPv6)
dbg: [net] Timeout while connecting to 44.135.218.201:3979
dbg: [net] Closed UDP listeners
dbg: [net] Initializing UDP listeners
dbg: [net] Closed UDP listeners
dbg: [net] Shutting down network
```

### Steps to reproduce

- Start game
- Open multiplayer

### PR DESCRIPTION
**Pull request by TrueBrain**
## Motivation / Problem

Fixes #9730.

## Description

As by commit message:
```
A race condition happens when an IPv6 connection takes more than
250ms to report an error, but does return before the IPv4 connection
is established.
In result, an invalid socket might be used for that connection.
```

Bit difficult story:
- You have an IPv4 and IPv6 connection, and are connecting to an IPv4+IPv6 server
- You try IPv6 first, but it takes more than 250ms to get a reply from the network stack
- Happy Eyeballs kicks in, and IPv4 connection is attempted too
- The IPv6 responds before the IPv4 does, and says: no connection

The code never assumed this situation to happen. As such, it didn't check if any of the remaining connections was actually connected, but instead assumed at least one of the remaining was. This in result means that `connected_socket` became `INVALID_SOCKET` and things went from bad to worse from there.

So now instead, check if any non-error'd socket is in writable state. If so, only than safely assume one of the sockets is connected.

The `if (this->sockets.empty()) {` meant this worked for most people, as it is rare two connections are actually pending, and the first reporting an error before the second manages to connect. Just when timing was bad, it broke.
(Normally IPv6 just times out, which takes a few seconds. The IPv4 attempt will succeed well before that, and that case would work perfectly fine).

The `if (this->sockets.empty()) {` can safely be removed, as the start of the function does exactly this already. So next `CheckActivity` the exact same code is executed and starts another connection attempt if needed. Less code duplication is beter :)

## Limitations

None

## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested')
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md)
* This PR affects the save game format? (label 'savegame upgrade')
* This PR affects the GS/AI API? (label 'needs review: Script API')
    * ai_changelog.hpp, gs_changelog.hpp need updating.
    * The compatibility wrappers (compat_*.nut) need updating.
* This PR affects the NewGRF API? (label 'needs review: NewGRF')
    * newgrf_debug_data.h may need updating.
    * [PR must be added to API tracker](https://wiki.openttd.org/en/Development/NewGRF/Specification%20Status)

### CODE REVIEW

File: src/network/core/tcp_connect.cpp [hunk 1]
```diff
@@ -371,18 +371,14 @@ bool TCPConnecter::CheckActivity()
 			closesocket(*it);
 			this->sock_to_address.erase(*it);
 			it = this->sockets.erase(it);
+			n--;
```
Comment (frosch123) [inline src/network/core/tcp_connect.cpp]: I don't understand this.
`n` is the number of bits set in `write_fd`. This loop does not check `write_fd`.
What says that the sockets with error are new errors?

Comment (TrueBrain) [reply src/network/core/tcp_connect.cpp]: There are only 2 possible ways that write_fd gets set: either the socket is connected or is in error. That is why this work. But also why I wrote the lines in "limitations" :)

Comment (frosch123) [reply src/network/core/tcp_connect.cpp]: Ah, I missed that the loop removed failed sockets from `sockets`. I was wondering whether there were still sockets which errored on the previous call.

**Review by frosch123 — DISMISSED** [review]

**Comment by RealDeuce** [comment]
Comfirmed that with this fix, issue #9730 did not happen 10/10 times.  It appears to be good. :shipit:

**Review by frosch123 — APPROVED** [review]

**Pull request merged**